fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SingleSidePredicate.java

```diff
@@ -0,0 +1,112 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.rewrite.logical;
+
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.rewrite.OneRewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Alias;
+import org.apache.doris.nereids.trees.expressions.EqualTo;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * push down expression which is not slot reference
+ */
+public class SingleSidePredicate extends OneRewriteRuleFactory {
+    /*
+     * rewrite example:
+     *       join(t1.a + 1 = t2.b + 2)                            join(c = d)
+     *             /       \                                         /    \
+     *            /         \                                       /      \
+     *           /           \             ====>                   /        \
+     *          /             \                                   /          \
+     *  olapScan(t1)     olapScan(t2)            project(t1.a + 1 as c)  project(t2.b + 2 as d)
+     *                                                        |                   |
+     *                                                        |                   |
+     *                                                        |                   |
+     *                                                        |                   |
+     *                                                     olapScan(t1)        olapScan(t2)
+     *TODO: now t1.a + t2.a = t1.b is not in hashJoinConjuncts. The rule will not handle it.
+     */
+    @Override
+    public Rule build() {
+        return logicalJoin()
+                .when(join -> join.getHashJoinConjuncts().stream().anyMatch(expr ->
+                        expr.children().stream().anyMatch(expr1 -> !(expr1 instanceof Slot))))
+                .then(join -> {
+                    List<List<Expression>> exprsOfJoinRelation =
+                            Lists.newArrayList(Lists.newArrayList(), Lists.newArrayList());
```

Comment:
a pair of list maybe better. BTW, `exprsOfJoinRelation` is not a good name

Reply:
no, if pair, more code will be written and can not use java stream